yt/frontends/cholla/data_structures.py

```diff
@@ -0,0 +1,165 @@
+        self.Parent = None
+        self.Children = []
+        self.Level = level
+        self.ActiveDimensions = dims
+
+    def __repr__(self):
+        return "ChollaGrid_%04i (%s)" % (self.id, self.ActiveDimensions)
+
+
+class ChollaHierarchy(GridIndex):
+    grid = ChollaGrid
+
+    def __init__(self, ds, dataset_type="cholla"):
+        self.dataset_type = dataset_type
+        self.dataset = weakref.proxy(ds)
+        # for now, the index file is the dataset!
+        self.index_filename = self.dataset.parameter_filename
+        self.directory = os.path.dirname(self.index_filename)
+        # float type for the simulation edges and must be float64 now
+        self.float_type = np.float64
+        super().__init__(ds, dataset_type)
+
+    def _detect_output_fields(self):
+        with h5py.File(self.index_filename, mode="r") as h5f:
+            self.field_list = [("cholla", k) for k in h5f.keys()]
+
+    def _count_grids(self):
+        self.num_grids = 1
+
+    def _parse_index(self):
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grid_particle_count[0][0] = 0
+        self.grid_levels[0][0] = 0
+        self.max_level = 0
+
+    def _populate_grid_objects(self):
+        self.grids = np.empty(self.num_grids, dtype="object")
+        for i in range(self.num_grids):
+            g = self.grid(i, self, self.grid_levels.flat[i], self.grid_dimensions[i])
+            g._prepare_grid()
+            g._setup_dx()
+            self.grids[i] = g
+
+
+class ChollaDataset(Dataset):
+    _index_class = ChollaHierarchy
+    _field_info_class = ChollaFieldInfo
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="cholla",
+        storage_filename=None,
+        units_override=None,
+        unit_system="cgs",
+    ):
+        self.fluid_types += ("cholla",)
+        super().__init__(filename, dataset_type, units_override=units_override)
+        self.storage_filename = storage_filename
+
+    def _set_code_unit_attributes(self):
+        # This is where quantities are created that represent the various
+        # on-disk units.  These are the defaults, but if they are listed
+        # in the HDF5 attributes for a file, which is loaded first, then those are
+        # used instead.
+        #
+        if not self.length_unit:
+            self.length_unit = self.quan(1.0, "pc")
+        if not self.mass_unit:
+            self.mass_unit = self.quan(1.0, "Msun")
+        if not self.time_unit:
+            self.time_unit = self.quan(1000, "yr")
+        if not self.velocity_unit:
+            self.velocity_unit = self.quan(1.0, "cm/s")
+        if not self.magnetic_unit:
+            self.magnetic_unit = self.quan(1.0, "gauss")
+
+        for key, unit in self.__class__.default_units.items():
+            setdefaultattr(self, key, self.quan(1, unit))
+
```

Comment:
Looks like all of this could be accomplished by just defining this class' `default_units`

Reply:
If we didn't allow the user to change the units, I see your that solution works, but I'm not sure that your suggestion will work while still checking each unit in the dataset attributes to see if the user overrides them.

Reply:
The `default_units` mechanism was primarily implemented for the AMRVAC frontend, where we very much allow units override. The key difference here is that AMRVAC data files never contain units, so I need to check if we need to generalize the mechanism to support this case. Maybe that's something better left done as a follow up refactor. I will think about it some more.

Reply:
Btw this is non blocking